### ISSUE TITLE
[pyQGIS] possible bug in raster editing

### ISSUE THREAD

**Issue by moovida** [qgis/QGIS#53060]
### What is the bug or the crash?

I am trying to use the raster API to create a new raster, starting from an existing raster layer.
To do so, I first create a new constant raster using a processing algorithm, and afterwards a QgsRasterBlock to edit single cells. 
There are oddities in the resulting raster values that might potentially be a bug.

### Steps to reproduce the issue

For the issue here I created a simple snippet that just copies cells from one raster to the other:

```
newRaster = processing.run("native:createconstantrasterlayer", {
    'EXTENT': dtmLayer.extent(),
    'TARGET_CRS': crs,
    'PIXEL_SIZE':dtmLayer.rasterUnitsPerPixelX(),
    'NUMBER':-9999.0,
    'OUTPUT_TYPE': 6,
    'OUTPUT':'TEMPORARY_OUTPUT'
    }
)

newRasterLayer = QgsRasterLayer(newRaster['OUTPUT'], 'temp', 'gdal')
newRasterProvider = newRasterLayer.dataProvider()

block = QgsRasterBlock(dataType, cols, rows)

for row in range(rows):
    for col in range(cols):
        point = dtmLayer.dataProvider().transformCoordinates(QgsPoint(col, row), transformType)
        value, res = dtmLayer.dataProvider().sample(QgsPointXY(point.x(), point.y()), 1)
        
        block.setValue(row, col, value)
         
newRasterProvider.setEditable(True)
newRasterProvider.writeBlock(block, band=1)
newRasterProvider.setEditable(False)
```

The dtmLayer can be any single band elevation model.

If using the info tool on the original and new raster, I find discrepancies between the values, that are supposed to be exactly the same:

![image](https://github.com/qgis/QGIS/assets/390250/a4f88baa-5ecf-4916-8d22-07d838f2f169)

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.30.2-'s-Hertogenbosch | QGIS code revision | 0992b53397
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.6
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.2 LTS
  |   |   |  
Active Python plugins
db_manager | 0.1.20
processing | 2.12.99
grassprovider | 2.12.99
MetaSearch | 0.3.6

</body></html>QGIS version
3.30.2-'s-Hertogenbosch
QGIS code revision
[0992b53397](https://github.com/qgis/QGIS/commit/0992b53397)
Qt version
5.15.3
Python version
3.10.6
GDAL/OGR version
3.4.1
PROJ version
8.2.1
EPSG Registry database version
v10.041 (2021-12-03)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.37.2
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.4
QScintilla2 version
2.11.6
OS version
Ubuntu 22.04.2 LTS

Active Python plugins
db_manager
0.1.20
processing
2.12.99
grassprovider
2.12.99
MetaSearch
0.3.6

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by nicogodet** [comment]
Best practice would be to ask on qgis user or dev mailing list before opening an issue.

Could you provide a standalone code snippet to reproduce ? Current requires definitions of variable: dtmLayer, crs, dataType, cols, rows, transformType.

**Comment by moovida** [comment]
Hi @nicogodet , sorry, I did ask in the mailinglist about something related and when preparing the data and snippet for the other issue, I stumbled into this.

I will prepare a standalone snippet right away.

**Comment by moovida** [comment]
Here the snippet and the raster I used:

```
from qgis import processing

dtm = "/path_to/dtm_flanginec.asc"
dtmLayer = QgsRasterLayer(dtm, "dtm")
# crs = dtmLayer.metadata().crs()   # crs is somehow not recognised
crs = QgsCoordinateReferenceSystem('EPSG:3003') # setting it manually

dataType = dtmLayer.dataProvider().dataType(1)
QgsProject.instance().addMapLayer(dtmLayer)

newRaster = processing.run("native:createconstantrasterlayer", {
    'EXTENT': dtmLayer.extent(),
    'TARGET_CRS': crs,
    'PIXEL_SIZE':dtmLayer.rasterUnitsPerPixelX(),
    'NUMBER':-9999.0,
    'OUTPUT_TYPE': 6,
    'OUTPUT':'TEMPORARY_OUTPUT'
    }
)
newRasterLayer = QgsRasterLayer(newRaster['OUTPUT'], 'temp', 'gdal')
newRasterProvider = newRasterLayer.dataProvider()

block = QgsRasterBlock(dataType, cols, rows)

for row in range(rows):
    for col in range(cols):
        point = dtmLayer.dataProvider().transformCoordinates(QgsPoint(col, row), transformType)
        value, res = dtmLayer.dataProvider().sample(QgsPointXY(point.x(), point.y()), 1)
        
        block.setValue(row, col, value)
            
newRasterProvider.setEditable(True)
newRasterProvider.writeBlock(block, band=1)
newRasterProvider.setEditable(False)

QgsProject.instance().addMapLayer(newRasterLayer)
```

Here the raster:

[dtm_flanginec.zip](https://github.com/qgis/QGIS/files/11439206/dtm_flanginec.zip)

**Comment by agiudiceandrea** [comment]
@moovida, the script is incomplete: cols and rows names are not defined.

**Comment by agiudiceandrea** [comment]
@moovida, transformType is also missing in the script (I suppose is value is 0).
Anyway it seems to me the issue is in the identify tool at little scale. Have you tried to zoom in to the identified point at a greater scale and use the identify tool again for the same coordinates? Have you tried to compare the two rasters using the raster calculator or to compare the two rasters blocks cell by cell?

**Comment by moovida** [comment]
> @moovida, transformType is also missing in the script (I suppose its value is 0). 

Yes exactly (was transformType = QgsRasterDataProvider.TransformType.TransformImageToLayer )

> Anyway it seems to me the issue is in the identify tool at little scale. Have you tried to zoom in to 
> the identified point at a greater scale and use the identify tool again for the same coordinates? 
> Have you tried to compare the two rasters using the raster calculator or to compare the two rasters blocks cell by cell? 

I tried to zoom in and it seems that you are right. It is the info tool that seems to be picking different coordinates even if the layers are supposed to be equal.

Then I tried to loop over both rasters and compare cell values. And that nailed the fact that you are right. The values are equal. Thanks. If I can close the issue, I will.

> Why don't you use the [QgsRasterDataProvider block()](https://qgis.org/pyqgis/3.28/core/QgsRasterDataProvider.html#qgis.core.QgsRasterDataProvider.block) 
> function to directly extract the data from the raster instead of transformCoordinates and sample?

Basically because I did not see that method. That is exactly what I was looking for, thanks :-)

**Comment by agiudiceandrea** [comment]
@moovida, could you please open another issue report about the identify tool?

**Comment by moovida** [comment]
Yes, sure. I was not sure if that was an expected/accepted behavior.
